### PR TITLE
Windows: Daemon default to Hyper-V containers on client

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

Fixes https://github.com/Microsoft/Virtualization-Documentation/issues/200. On client SKUs, sets the default isolation to Hyper-V and errors out if set to process, as client does not support Windows server containers. No change to server SKUs.

@mebersol @taylorb-microsoft

### CODE REVIEW

**Comment by taylorb-microsoft** [comment]
We haven't even officially started supporting Win10 yet and this issue is already a top hit... This should help a lot!

**Comment by justincormack** [comment]
Looks good.

**Comment by jhowardmsft** [comment]
@thaJeztah Can this be moved to code review? Thx.

**Comment by thaJeztah** [comment]
@jhowardmsft oh! Yes, I think it can

File: daemon/daemon_windows.go [hunk 1]
```diff
@@ -485,6 +489,12 @@ func (daemon *Daemon) setDefaultIsolation() error {
 			if containertypes.Isolation(val).IsHyperV() {
 				daemon.defaultIsolation = containertypes.Isolation("hyperv")
 			}
+			if containertypes.Isolation(val).IsProcess() {
```
Comment (thaJeztah) [inline daemon/daemon_windows.go]: Shouldn't this be moved to `IsValid()`? IIUC for this type of daemon, "process" isolation simply isn't a valid option

File: daemon/daemon_windows.go [hunk 2]
```diff
@@ -470,6 +470,10 @@ func (daemon *Daemon) stats(c *container.Container) (*types.StatsJSON, error) {
 // daemon to run in. This is only applicable on Windows
 func (daemon *Daemon) setDefaultIsolation() error {
 	daemon.defaultIsolation = containertypes.Isolation("process")
+	// On client SKUs, default to Hyper-V
+	if system.IsWindowsClient() {
+		daemon.defaultIsolation = containertypes.Isolation("hyperv")
```
Comment (thaJeztah) [inline daemon/daemon_windows.go]: Perhaps change this to if/else?

**Comment by thaJeztah** [comment]
Can you also update the documentation around this setting? Describe that the default is different for client/server, and that client doesn't support "process";

- https://github.com/docker/docker/blob/master/docs/reference/commandline/dockerd.md#options-for-the-runtime
- https://github.com/docker/docker/blob/master/docs/reference/commandline/run.md#specify-isolation-technology-for-container---isolation
- https://github.com/docker/docker/blob/master/docs/reference/commandline/create.md#specify-isolation-technology-for-container---isolation
- https://github.com/docker/docker/blob/master/man/docker-create.1.md#specify-isolation-technology-for-container---isolation
- https://github.com/docker/docker/blob/master/man/docker-run.1.md#specify-isolation-technology-for-container---isolation
- (section seems to be missing in the dockerd man-page https://github.com/docker/docker/blob/master/man/dockerd.8.md)

File: daemon/daemon_windows.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/daemon_windows.go]: TBH both are equally valid. I prefer it here as strictly the option chosen is a valid recognised option for isolation which is what that function is checking. I don't mind either way though, just as I say prefer it here.

File: daemon/daemon_windows.go (hunk 2 above)
Comment (jhowardmsft) [inline daemon/daemon_windows.go]: @LK4D4 would comment that I should flatten it it if I did :smile:

**Comment by jhowardmsft** [comment]
Docs updated. @neilpeterson FYI for matching Microsoft side docs.

File: daemon/daemon_windows.go (hunk 1 above)
Comment (thaJeztah) [inline daemon/daemon_windows.go]: Alright, possibly just a matter of preference 👍

File: daemon/daemon_windows.go (hunk 2 above)
Comment (thaJeztah) [inline daemon/daemon_windows.go]: Heh, I think @LK4D4 usually doesn't like "else" if the first branch does a *return*; in this case it's assigning a value twice (overriding the first), so it felt cleaner to do if/else, but it's just a nit

**Comment by thaJeztah** [comment]
LGTM, thanks!

**Comment by cpuguy83** [comment]
Ok, so there will be a docker daemon running on Windows 10 (client) which will only work with hyper-v, is that right?

**Comment by jhowardmsft** [comment]
Yes.

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**